zio-k8s-operator/src/main/scala/com/coralogix/zio/k8s/operator/leader/locks/LeaseLock.scala

```diff
@@ -0,0 +1,229 @@
+                 .acquire()
+                 .provideSome[Clock with Logging](_ ++ Has(random))
+                 .toManaged(_ => impl.release())
+      _     <- impl.renew().fork.toManaged_
+    } yield ()
+
+  class Impl(
+    store: Ref[Option[VersionedRecord]],
+    namespace: K8sNamespace,
+    identity: String
+  ) {
+    def acquire(): ZIO[Random with Clock with Logging, LeaderElectionFailure[Nothing], Unit] =
+      tryAcquireOrRenew().asSomeError
+        .flatMap {
+          case false => ZIO.fail(None)
+          case true  => ZIO.unit
+        }
+        .zipLeft(log.info("Leadership acquired"))
+        .tapError {
+          case None          => ZIO.unit
+          case Some(failure) =>
+            logLeaderElectionFailure(failure) *>
+              log.info("Failed to take leadership, will retry...")
+        }
+        .retry(Schedule.fixed(retryPeriod).jittered(0.0, 1.2))
+        .optional
+        .unit
+
+    def renew(): ZIO[Clock with Logging, LeaderElectionFailure[Nothing], Unit] = {
+      tryAcquireOrRenew()
+        .retry(
+          (Schedule.fixed(retryPeriod) >>> Schedule.elapsed)
+            .whileOutput(_ < renewTimeout)
+            .as(false)
+        ) // trying to renew within the renew timeout period
+        .repeat(
+          Schedule.fixed(retryPeriod).whileInput((success: Boolean) => success)
+        ) // keep renewing if it succeeded
+    } *> log.info("Leadership lost") *> leadershipLost.offer(()).unit
+
+    def release(): ZIO[Clock with Logging, Nothing, Unit] =
+      store.get
+        .flatMap {
+          case Some(stored) if stored.record.holderIdentity == identity =>
+            for {
+              now      <- clock.currentDateTime.mapError(DateTimeError)
+              newRecord =
+                LeaderElectionRecord(
+                  holderIdentity = identity,
+                  leaseDuration = 1.second,
+                  now,
+                  now,
+                  stored.record.leaderTransitions
+                )
+              _        <- update(newRecord, stored.resourceVersion)
+            } yield ()
+          case _                                                        => ZIO.unit
+        }
+        .catchAll { error =>
+          logLeaderElectionFailure(error)
+        }
+
+    private def tryAcquireOrRenew()
+      : ZIO[Clock with Logging, LeaderElectionFailure[Nothing], Boolean] =
+      for {
+        latest <- get()
+        now    <- clock.currentDateTime.mapError(DateTimeError)
+        result <- latest match {
+                    case None            =>
+                      val record = LeaderElectionRecord(identity, leaseDuration, now, now, 0)
+                      for {
+                        _               <-
+                          log.info(
+                            s"Lease record $lockName does not exist, creating and taking leadership"
+                          )
+                        resourceVersion <- create(record)
```

Comment:
Will this fail if someone else acquired the lock between this and the call to `get`?

Reply:
Also important: what happens if we are interrupted right after this call? e.g. if the JVM is shutdown. Will Kubernetes delete the lock?

Reply:
Note that this whole algorithm does not guarantee exclusive leadership in all corner cases :( 

About the ones you mentioned:
- If the lease did not exist and multiple pods are racing to create it first, the first create will win and the others are causing a `K8sFailure` which is handled and triggers the retry of the acquisition. So I think there are no problems with this one.
- The lease resource is never deleted by either k8s or the library. I think probably how we would use it is to pre-create the lease with the deployment. If the current leader gets interrupted in any way then the other pods that are continuously trying to acquire the lock will eventually (within the lease duration) win. The only way to make it stuck is to write some very large (future) `renewTime` in the resource.

Reply:
Right. I would also consider two additions:
* If you are a leader, delete the node when you are interrupted, assuming that the delete API call can use a resourceVersion so we are sure that we don't delete someone else's node;
* Have the other pods watch the resource to speed up the process of electing a new leader when the current one steps down

Reply:
What other cases have you noted in which we don't guarantee exclusive leadership? That's unfortunate. We could try and fix them.

Reply:
Instead of deleting, there is a release operation that sets the renew time to very close so others can immediately pick it up (thats what is implemented in go and fabric8). Maybe better than deletion so the lease keeps history (at least an increasing resource version)?

Implementing it with watch sounds like a good idea that I did not consider because of just blindly trying to parse the existing go implementation ;)

I think what can happen is that:
- With some clock skew it is possible that a non-leader steals the ownership. Normally this does not happen because the leader frequently updates the lease (default options is that in every 2 sec it updates it and the lease time is 15sec). 
- If someone stole the leadership then I'm interrupting the leader-election protected effect
- But this means that for a short time there are two leaders

Reply:
Running on Kubernetes basically guarantees that you won't see major clock skew as all Kubernetes distributions more or less enforce having NTP running. So if clock skew is the problematic factor I think we can disregard it